# S05 T01: Tasca mètodes de mostreig
**Descripció**: Aprèn a realitzar mostreig de les dades amb Python.

## Exercici 1
**Agafa un conjunt de dades de tema esportiu que t'agradi. Realitza un mostreig de les dades generant una mostra aleatòria simple i una mostra sistemàtica.**

Sample Method Description: 

https://towardsdatascience.com/the-5-sampling-algorithms-every-data-scientist-need-to-know-43c7bc11d17c

https://www.scribbr.com/methodology/sampling-methods/

https://www.analyticsvidhya.com/blog/2019/09/data-scientists-guide-8-types-of-sampling-techniques/

In [105]:
import numpy as np
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt

We import an NBA player dataset.

https://www.kaggle.com/nathanlauga/nba-games?select=games_details.csv

In [106]:
df_NBA = pd.read_csv('NBAgames_details.csv')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (6) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [107]:
df_NBA.head()

GAME_ID     TEAM_ID TEAM_ABBREVIATION   TEAM_CITY  PLAYER_ID  \
0  22100213  1610612764               WAS  Washington     203484   
1  22100213  1610612764               WAS  Washington    1628398   
2  22100213  1610612764               WAS  Washington    1629655   
3  22100213  1610612764               WAS  Washington     203078   
4  22100213  1610612764               WAS  Washington     203915   

                PLAYER_NAME    NICKNAME START_POSITION COMMENT    MIN  ...  \
0  Kentavious Caldwell-Pope  Kentavious              F     NaN  27:41  ...   
1                Kyle Kuzma        Kyle              F     NaN  30:28  ...   
2            Daniel Gafford      Daniel              C     NaN  24:21  ...   
3              Bradley Beal     Bradley              G     NaN  35:07  ...   
4         Spencer Dinwiddie     Spencer              G     NaN  28:34  ...   

   OREB  DREB  REB  AST  STL  BLK   TO   PF   PTS  PLUS_MINUS  
0   1.0   5.0  6.0  2.0  1.0  0.0  1.0  0.0   3.0         2.0  
1   1.0   4.0  5.0  3.0  1.0  2.0  1.0  1.0   5.0       -14.0  
2   2.0   7.0  9.0  1.0  2.0  1.0  1.0  4.0  20.0        -2.0  
3   0.0   3.0  3.0  7.0  2.0  0.0  2.0  3.0  24.0        -9.0  
4   0.0   3.0  3.0  2.0  0.0  0.0  2.0  1.0   0.0        -5.0  

[5 rows x 29 columns]

In [108]:
df_NBA.TEAM_CITY.value_counts()

Los Angeles                  35326
Miami                        22452
San Antonio                  22239
Boston                       22078
Dallas                       21470
Cleveland                    21240
Indiana                      21223
Denver                       21134
Detroit                      21103
Golden State                 21080
Atlanta                      20981
Utah                         20967
Houston                      20949
Milwaukee                    20923
Toronto                      20895
Portland                     20733
Phoenix                      20728
Memphis                      20692
Orlando                      20553
Washington                   20548
Chicago                      20489
Philadelphia                 20384
Sacramento                   19936
New York                     19858
Minnesota                    19773
Charlotte                    19046
New Orleans                  18317
Oklahoma City                15673
Brooklyn            

In [109]:
df_NBA.columns

Index(['GAME_ID', 'TEAM_ID', 'TEAM_ABBREVIATION', 'TEAM_CITY', 'PLAYER_ID',
       'PLAYER_NAME', 'NICKNAME', 'START_POSITION', 'COMMENT', 'MIN', 'FGM',
       'FGA', 'FG_PCT', 'FG3M', 'FG3A', 'FG3_PCT', 'FTM', 'FTA', 'FT_PCT',
       'OREB', 'DREB', 'REB', 'AST', 'STL', 'BLK', 'TO', 'PF', 'PTS',
       'PLUS_MINUS'],
      dtype='object')

Definition of some columns is:
 * MIN: minutes
 * FGM, FGA, FG_PCT: field goals made, attempted and percentage
 * FG3M, FG3A, FG3_PCT: three-point field goals made, attempted and percentage
 * FTM, FTA, FT_PCT: free throws made, attempted and percentage
 * REB, OREB, DREB: rebounds, offensive rebounds, defensive rebounds
 * AST: assists
 * STL: steals
 * BLK: blocks
 * TO: turnovers
 * PF: personal fouls
 * PTS: points
 * PLUS_MINUS: The point differential when a player or team is on the floor

In [110]:
df_NBA.shape

(626111, 29)

**Simple Random Sampling**

We ramdomly select 1000 of the 626111 players

In [111]:
df_sample=df_NBA.sample(1000)

In [112]:
df_sample.head()

GAME_ID     TEAM_ID TEAM_ABBREVIATION   TEAM_CITY  PLAYER_ID  \
117830  20700510  1610612737               ATL     Atlanta       1533   
482613  21501182  1610612764               WAS  Washington     201196   
222147  20400437  1610612749               MIL   Milwaukee        693   
24887   22000028  1610612750               MIN   Minnesota    1629633   
497357  21500609  1610612746               LAC          LA     203143   

            PLAYER_NAME NICKNAME START_POSITION COMMENT    MIN  ...  OREB  \
117830  Anthony Johnson      NaN              G     NaN  28:02  ...   1.0   
482613   Ramon Sessions      NaN              G     NaN  32:17  ...   0.0   
222147        Joe Smith      NaN              F     NaN  33:26  ...   2.0   
24887    Jarrett Culver      NaN            NaN     NaN  19:45  ...   3.0   
497357   Pablo Prigioni      NaN            NaN     NaN  16:11  ...   2.0   

        DREB  REB  AST  STL  BLK   TO   PF   PTS  PLUS_MINUS  
117830   0.0  1.0  6.0  1.0  0.0  0.0  1.0   5.0         5.0  
482613   1.0  1.0  6.0  1.0  0.0  2.0  2.0  12.0        -9.0  
222147   4.0  6.0  1.0  0.0  0.0  1.0  4.0   6.0       -16.0  
24887    3.0  6.0  0.0  1.0  0.0  0.0  3.0  14.0         9.0  
497357   2.0  4.0  4.0  0.0  0.0  1.0  0.0   5.0         0.0  

[5 rows x 29 columns]

In [113]:
df_sample.shape

(1000, 29)

**Systematic Sampling**

In [114]:
# Define systematic sampling function
def systematic_sampling(df, step):
    initial_value = np.random.randint(0,111)
    indexes = np.arange(initial_value, len(df), step=step)
    systematic_sample = df.iloc[indexes]
    return systematic_sample
 
 
# Obtain a systematic sample and save it in a new variable
df_syst_sample = systematic_sampling(df_NBA, 625) # 625 is the step of the systematic sampling. We select it to have around 1000 players

In [115]:
df_syst_sample

GAME_ID     TEAM_ID TEAM_ABBREVIATION     TEAM_CITY  PLAYER_ID  \
21      22100213  1610612766               CHA     Charlotte    1629667   
646     22100209  1610612747               LAL   Los Angeles     201566   
1271    22100183  1610612744               GSW  Golden State    1626172   
1896    22100150  1610612742               DAL        Dallas    1628425   
2521    22100129  1610612750               MIN     Minnesota     201976   
...          ...         ...               ...           ...        ...   
623146  21200014  1610612759               SAS   San Antonio     201986   
623771  11200102  1610612750               MIN     Minnesota     200826   
624396  11200078  1610612737               ATL       Atlanta     202345   
625021  11200055  1610612766               CHA     Charlotte     201196   
625646  11200026  1610612755               PHI  Philadelphia     201950   

              PLAYER_NAME  NICKNAME START_POSITION                 COMMENT  \
21        Jalen McDaniels     Jalen            NaN                     NaN   
646     Russell Westbrook   Russell              G                     NaN   
1271         Kevon Looney     Kevon              C                     NaN   
1896       Sterling Brown  Sterling            NaN                     NaN   
2521     Patrick Beverley   Patrick              G                     NaN   
...                   ...       ...            ...                     ...   
623146      Nando De Colo       NaN            NaN  DNP - Coach's Decision   
623771         J.J. Barea       NaN            NaN                     NaN   
624396       Damion James       NaN            NaN                     NaN   
625021     Ramon Sessions       NaN            NaN                     NaN   
625646       Jrue Holiday       NaN            NaN                     NaN   

          MIN  ...  OREB  DREB   REB  AST  STL  BLK   TO   PF   PTS  \
21      17:19  ...   0.0   6.0   6.0  1.0  0.0  0.0  0.0  1.0  11.0   
646     31:15  ...   0.0   6.0   6.0  8.0  1.0  0.0  4.0  3.0  25.0   
1271    20:27  ...   3.0   7.0  10.0  1.0  0.0  0.0  0.0  0.0   4.0   
1896    11:13  ...   0.0   4.0   4.0  0.0  0.0  0.0  0.0  1.0   0.0   
2521    27:21  ...   0.0   4.0   4.0  3.0  0.0  0.0  2.0  6.0   5.0   
...       ...  ...   ...   ...   ...  ...  ...  ...  ...  ...   ...   
623146    NaN  ...   NaN   NaN   NaN  NaN  NaN  NaN  NaN  NaN   NaN   
623771     18  ...   0.0   0.0   0.0  7.0  0.0  0.0  1.0  1.0   2.0   
624396     16  ...   2.0   5.0   7.0  1.0  1.0  1.0  1.0  5.0  10.0   
625021     34  ...   1.0   3.0   4.0  7.0  1.0  0.0  3.0  2.0  18.0   
625646     20  ...   2.0   2.0   4.0  3.0  1.0  0.0  1.0  3.0  27.0   

        PLUS_MINUS  
21             8.0  
646          -13.0  
1271          14.0  
1896           1.0  
2521         -23.0  
...            ...  
623146         NaN  
623771         NaN  
624396         NaN  
625021         NaN  
625646         NaN  

[1002 rows x 29 columns]

## Exercici 2
**Continua amb el conjunt de dades de tema esportiu i genera una mostra estratificada i una mostra utilitzant SMOTE (Synthetic Minority Oversampling Technique).**

**Stratified Sampling**

1000 players are selected proportionally to the number or times the TEAM_CITY 

In [116]:
from sklearn.model_selection import train_test_split

In [117]:
x = df_NBA
y = df_NBA.TEAM_CITY
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y, train_size=1000)

In [118]:
x_train

GAME_ID     TEAM_ID TEAM_ABBREVIATION     TEAM_CITY  PLAYER_ID  \
182071  20500667  1610612765               DET       Detroit       2545   
84822   20800463  1610612742               DAL        Dallas        467   
444427  21601222  1610612751               BKN      Brooklyn    1627785   
232395  20400015  1610612746               LAC   Los Angeles       2733   
390637  21800529  1610612739               CLE     Cleveland    1626224   
...          ...         ...               ...           ...        ...   
389547  21800578  1610612755               PHI  Philadelphia    1628413   
226385  20400262  1610612758               SAC    Sacramento       2440   
526368  21400922  1610612742               DAL        Dallas     200765   
157668  20600258  1610612756               PHX       Phoenix        147   
364064  21900177  1610612752               NYK      New York    1628422   

             PLAYER_NAME NICKNAME START_POSITION COMMENT    MIN  ...  OREB  \
182071     Darko Milicic      NaN            NaN     NaN   3:37  ...   0.0   
84822         Jason Kidd      NaN              G     NaN  39:30  ...   3.0   
444427  Isaiah Whitehead      NaN            NaN     NaN  25:52  ...   0.0   
232395  Shaun Livingston      NaN            NaN     NaN  16:04  ...   0.0   
390637        Cedi Osman      NaN              F     NaN  35:42  ...   2.0   
...                  ...      ...            ...     ...    ...  ...   ...   
389547      Jonah Bolden      NaN              F     NaN  29:03  ...   3.0   
226385       Matt Barnes      NaN            NaN     NaN   8:57  ...   1.0   
526368       Rajon Rondo      NaN              G     NaN  31:54  ...   0.0   
157668        Jalen Rose      NaN            NaN     NaN   6:30  ...   0.0   
364064    Damyean Dotson      NaN            NaN     NaN  22:49  ...   0.0   

        DREB  REB   AST  STL  BLK   TO   PF   PTS  PLUS_MINUS  
182071   0.0  0.0   0.0  0.0  0.0  0.0  0.0   2.0        -4.0  
84822    4.0  7.0  16.0  4.0  1.0  2.0  0.0  14.0        10.0  
444427   3.0  3.0   2.0  0.0  0.0  3.0  3.0   5.0       -22.0  
232395   5.0  5.0   3.0  0.0  0.0  1.0  2.0   6.0         5.0  
390637   6.0  8.0   3.0  0.0  0.0  3.0  5.0  22.0         8.0  
...      ...  ...   ...  ...  ...  ...  ...   ...         ...  
389547   6.0  9.0   3.0  0.0  2.0  1.0  4.0  11.0        11.0  
226385   0.0  1.0   0.0  0.0  0.0  0.0  0.0   3.0         7.0  
526368   4.0  4.0   6.0  2.0  0.0  0.0  3.0  14.0       -21.0  
157668   1.0  1.0   1.0  0.0  0.0  0.0  0.0   3.0         2.0  
364064   3.0  3.0   2.0  0.0  0.0  0.0  1.0   4.0         3.0  

[1000 rows x 29 columns]

**SMOTE Sample (Synthetic Minority Oversampling Technique)**

We are going to check the players that have play and they one that haven´t and use the SMOTE to simulate addional samples and avoid high differences in the variable weight.

In [119]:
# !pip install imbalanced-learn

In [120]:
from imblearn.over_sampling import SMOTE
from collections import Counter

SMOTE will only work with numerical values, so we have to clean up the dataset.

For this exercise we keep it simple and we remove string columns, fill all nulls and crease a column with a 1 if the players have played any minute or with a 0 if they haven´t played any minute.

In [121]:
df_NBA['MIN'].fillna(0, inplace=True)
df_NBA.MIN.value_counts()

0        102360
5          1989
12:00      1654
24         1185
22         1123
          ...  
57:50         1
56:29         1
55:28         1
51:37         1
54:48         1
Name: MIN, Length: 3390, dtype: int64

In [122]:
df_NBA['played']=1
df_NBA.loc[(df_NBA['MIN']==0), 'played']=0
df_NBA['played'].value_counts()           

1    523751
0    102360
Name: played, dtype: int64

In [123]:
# We remove string columns and fill all NaN with 0
df_NBA2 = df_NBA.copy()
df_NBA2.drop(['GAME_ID', 'TEAM_ID', 'TEAM_ABBREVIATION', 'TEAM_CITY', 'PLAYER_ID',
       'PLAYER_NAME', 'NICKNAME', 'START_POSITION', 'COMMENT', 'MIN'], axis=1, inplace=True)
df_NBA2.fillna(0, inplace=True)

In [124]:
x = df_NBA2.drop('played', axis=1) #x is the NBA data set with only numeric values but the played column
y = df_NBA2.played
print('Original dataset shape %s' % Counter(y))

Original dataset shape Counter({1: 523751, 0: 102360})


523751 players have played some minute and 102360 players have NOT played any minute. There is a big difference in % so we will use SMOTE to simulate further players that don´t have played

In [125]:
x

FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  DREB  \
0       1.0   6.0   0.167   0.0   5.0    0.000  1.0  1.0   1.000   1.0   5.0   
1       2.0  12.0   0.167   1.0   8.0    0.125  0.0  0.0   0.000   1.0   4.0   
2       9.0  12.0   0.750   0.0   0.0    0.000  2.0  5.0   0.400   2.0   7.0   
3       9.0  20.0   0.450   5.0  11.0    0.455  1.0  1.0   1.000   0.0   3.0   
4       0.0   5.0   0.000   0.0   4.0    0.000  0.0  0.0   0.000   0.0   3.0   
...     ...   ...     ...   ...   ...      ...  ...  ...     ...   ...   ...   
626106  4.0   9.0   0.444   3.0   6.0    0.500  6.0  7.0   0.857   0.0   2.0   
626107  7.0  11.0   0.636   0.0   0.0    0.000  4.0  4.0   1.000   1.0   0.0   
626108  3.0   7.0   0.429   0.0   0.0    0.000  0.0  0.0   0.000   3.0   5.0   
626109  1.0   1.0   1.000   0.0   0.0    0.000  0.0  2.0   0.000   1.0   2.0   
626110  3.0   6.0   0.500   0.0   0.0    0.000  2.0  4.0   0.500   0.0   2.0   

        REB  AST  STL  BLK   TO   PF   PTS  PLUS_MINUS  
0       6.0  2.0  1.0  0.0  1.0  0.0   3.0         2.0  
1       5.0  3.0  1.0  2.0  1.0  1.0   5.0       -14.0  
2       9.0  1.0  2.0  1.0  1.0  4.0  20.0        -2.0  
3       3.0  7.0  2.0  0.0  2.0  3.0  24.0        -9.0  
4       3.0  2.0  0.0  0.0  2.0  1.0   0.0        -5.0  
...     ...  ...  ...  ...  ...  ...   ...         ...  
626106  2.0  0.0  2.0  0.0  1.0  3.0  17.0         0.0  
626107  1.0  1.0  1.0  0.0  3.0  3.0  18.0         0.0  
626108  8.0  0.0  1.0  0.0  0.0  3.0   6.0         0.0  
626109  3.0  1.0  0.0  0.0  4.0  2.0   2.0         0.0  
626110  2.0  6.0  2.0  0.0  6.0  1.0   8.0         0.0  

[626111 rows x 19 columns]

In [126]:
smote = SMOTE()
x_sm, y_sm = smote.fit_resample(x, y)
print('Original dataset shape %s' % Counter(y_sm))

Original dataset shape Counter({1: 523751, 0: 523751})


The number of players that have played or NOT any minutes is now the same

In [127]:
x_sm

FGM   FGA  FG_PCT  FG3M  FG3A  FG3_PCT  FTM  FTA  FT_PCT  OREB  DREB  \
0        1.0   6.0   0.167   0.0   5.0    0.000  1.0  1.0     1.0   1.0   5.0   
1        2.0  12.0   0.167   1.0   8.0    0.125  0.0  0.0     0.0   1.0   4.0   
2        9.0  12.0   0.750   0.0   0.0    0.000  2.0  5.0     0.4   2.0   7.0   
3        9.0  20.0   0.450   5.0  11.0    0.455  1.0  1.0     1.0   0.0   3.0   
4        0.0   5.0   0.000   0.0   4.0    0.000  0.0  0.0     0.0   0.0   3.0   
...      ...   ...     ...   ...   ...      ...  ...  ...     ...   ...   ...   
1047497  0.0   0.0   0.000   0.0   0.0    0.000  0.0  0.0     0.0   0.0   0.0   
1047498  0.0   0.0   0.000   0.0   0.0    0.000  0.0  0.0     0.0   0.0   0.0   
1047499  0.0   0.0   0.000   0.0   0.0    0.000  0.0  0.0     0.0   0.0   0.0   
1047500  0.0   0.0   0.000   0.0   0.0    0.000  0.0  0.0     0.0   0.0   0.0   
1047501  0.0   0.0   0.000   0.0   0.0    0.000  0.0  0.0     0.0   0.0   0.0   

         REB  AST  STL  BLK   TO   PF   PTS  PLUS_MINUS  
0        6.0  2.0  1.0  0.0  1.0  0.0   3.0         2.0  
1        5.0  3.0  1.0  2.0  1.0  1.0   5.0       -14.0  
2        9.0  1.0  2.0  1.0  1.0  4.0  20.0        -2.0  
3        3.0  7.0  2.0  0.0  2.0  3.0  24.0        -9.0  
4        3.0  2.0  0.0  0.0  2.0  1.0   0.0        -5.0  
...      ...  ...  ...  ...  ...  ...   ...         ...  
1047497  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0  
1047498  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0  
1047499  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0  
1047500  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0  
1047501  0.0  0.0  0.0  0.0  0.0  0.0   0.0         0.0  

[1047502 rows x 19 columns]

Usefull links about SMOTE:

https://machinelearningmastery.com/smote-oversampling-for-imbalanced-classification/

https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOTE.html

https://towardsdatascience.com/how-to-effortlessly-handle-class-imbalance-with-python-and-smote-9b715ca8e5a7


## Exercici 3
**Continua amb el conjunt de dades de tema esportiu i genera una mostra utilitzant el mètode Reservoir sampling.**

I love this problem statement:

Say you have a stream of items of large and unknown length that we can only iterate over once.
Create an algorithm that randomly chooses an item from this stream such that each item is equally likely to be selected.

We are going to select randomly 1000 players from our NBA Dataset

In [128]:
import random

In [129]:
k=1000
reservoir = []
for i, element in enumerate(df_NBA.PLAYER_ID):
    if i+1<= k:
        reservoir.append(element)
    else:
        probability = k/(i+1)
        if random.random() < probability:
            # Select item in df_NBA and remove one of the k items already selected
             reservoir[random.choice(range(0,k))] = element

In [130]:
reservoir

[204038,
 2553,
 203095,
 202684,
 201588,
 201577,
 2427,
 2617,
 101155,
 201229,
 1629027,
 202682,
 1752,
 1731,
 966,
 203094,
 201196,
 202684,
 202684,
 2403,
 1723,
 1629667,
 202683,
 201943,
 2592,
 202731,
 203471,
 1629631,
 1502,
 201238,
 201568,
 1626143,
 201177,
 203493,
 2617,
 397,
 200797,
 1065,
 2754,
 101145,
 2205,
 203917,
 202087,
 201188,
 2034,
 203935,
 203526,
 1629633,
 201858,
 1730,
 1627732,
 101133,
 203084,
 202334,
 2222,
 202721,
 358,
 201965,
 2563,
 202688,
 1548,
 1729,
 101179,
 1882,
 202689,
 202337,
 101130,
 1909,
 101114,
 201977,
 203141,
 202325,
 201590,
 2250,
 201043,
 202335,
 203585,
 202708,
 1626159,
 201575,
 201952,
 2419,
 1629643,
 201563,
 101130,
 200809,
 2410,
 203909,
 243,
 1497,
 1628993,
 686,
 1629013,
 2758,
 201587,
 202718,
 1960,
 2549,
 101130,
 201939,
 2200,
 203496,
 203481,
 2557,
 204222,
 959,
 1628391,
 203103,
 2200,
 202732,
 1628972,
 201192,
 101187,
 201579,
 1628510,
 1627851,
 283,
 2544,
 202694,


In [131]:
len(reservoir)

1000